heat/core/manipulations.py

```diff
@@ -2798,146 +2922,96 @@ def unique(
-        # find indices of vectors
-        if a.comm.Get_rank() == max_pos.item():
-            # Get indices of the unique vectors to share with all over processes
-            indices = inverse_pos.reshape(-1).unique()
-        else:
-            indices = torch.empty((max_uniques.item(),), dtype=inverse_pos.dtype)
-
-        a.comm.Bcast(indices, root=max_pos)
-
-        gres = local_data[indices.tolist()]
-
-        inverse_indices = indices
-        if sorted:
-            raise ValueError(
-                "Sorting with axis != split is not supported yet. "
-                "See https://github.com/helmholtz-analytics/heat/issues/363"
-            )
-
-    if axis is not None:
-        # transpose matrix back
-        gres = gres.transpose(0, axis)
+        # balance gres if needed
+        gres.balance_()
+        # global sorted unique
+        lres = __pivot_sorting(gres, torch.unique, 0, sorted=True, return_inverse=True)
+        # second local unique
+        if 0 not in lres.shape:
+            lres = torch.unique(lres, sorted=True, dim=unique_axis)
+        lres_split = 0
+
+    gres = factories.array(lres, dtype=a.dtype, is_split=lres_split, device=a.device)
+    gres.balance_()
 
-    split = split if a.split < len(gres.shape) else None
-    result = factories.array(
-        gres, dtype=a.dtype, device=a.device, comm=a.comm, split=split, is_split=is_split
-    )
-    if split is not None:
-        result.resplit_(a.split)
+    if return_inverse:
+        # inverse indices
+        # allocate local tensors and global DNDarray
+        inverse = torch.empty(inv_shape, dtype=torch.int64, device=local_data.device)
+        if a.is_distributed():
+            inv_split = 0 if inverse.ndim == 1 else a.split
+        else:
+            inv_split = None
+        global_inverse = factories.array(inverse, is_split=inv_split, device=gres.device)
+
+        unique_ranks = size if gres.is_distributed() else 1
+        if unique_ranks > 1:
+            gres_map = gres.create_lshape_map()
+            gres_offsets = torch.cat(
+                (torch.tensor([0], device=gres_map.device), gres_map[:-1, gres.split])
+            ).cumsum(dim=0)
+        else:
+            gres_map = torch.tensor(gres.gshape, device=inverse.device)
+            gres_offsets = torch.tensor([0], device=gres_map.device)
+        lres = gres.larray
+        for p in range(unique_ranks):
+            if unique_ranks == 1:
+                incoming_offset = 0
+            else:
+                origin = (rank - p) % size
+                incoming_offset = gres_offsets[origin]
+                tmp = torch.empty(
+                    gres_map[0].tolist(), dtype=local_data.dtype, device=local_data.device
+                )
+            # loop through unique elements, find matching position in data
+            for i, el in enumerate(lres):
+                counts = torch.zeros_like(local_data, dtype=torch.int8, device=local_data.device)
```

Comment:
is int8 too small here? can we spare the space for a larger dtype? i think this only goes to 128, which should be fine for a while, but int16/32 shouldnt cost that much

Reply:
you're right, I started this off as an array containing only 1s and 0s, but that evolved. Great you noticed, now int32